src/main/java/net/minestom/server/entity/Entity.java

```diff
@@ -311,33 +312,40 @@ public <TMeta extends EntityMeta> void editEntityMeta(Class<TMeta> metaClass, Co
      * @param chunks   the chunk indexes to load before teleporting the entity,
      *                 indexes are from {@link ChunkUtils#getChunkIndex(int, int)},
      *                 can be null or empty to only load the chunk at {@code position}
+     * @param flags    flags used to teleport the entity relatively rather than absolutely
      * @throws IllegalStateException if you try to teleport an entity before settings its instance
      */
-    public @NotNull CompletableFuture<Void> teleport(@NotNull Pos position, long @Nullable [] chunks) {
+    public @NotNull CompletableFuture<Void> teleport(@NotNull Pos position, long @Nullable [] chunks, @NotNull RelativeFlag... flags) {
         Check.stateCondition(instance == null, "You need to use Entity#setInstance before teleporting an entity!");
+        int relativeBits = RelativeFlag.getBitsFromFlags(flags);
+        final Pos globalPosition = PositionUtils.getPositionWithRelativeFlags(this.position, position, relativeBits);
         final Runnable endCallback = () -> {
             this.previousPosition = this.position;
-            this.position = position;
-            refreshCoordinate(position);
-            synchronizePosition(true);
-            sendPacketToViewers(new EntityHeadLookPacket(getEntityId(), position.yaw()));
+            this.position = globalPosition;
+            refreshCoordinate(globalPosition);
+            if (this instanceof Player player) player.synchronizePositionAfterTeleport(position, relativeBits);
```

Comment:
This feels worse than how it was before. There shouldn't be player specific logic in entity if it can be handled by an override in Player (which keeps it self contained there)

Reply:
> This feels worse than how it was before. There shouldn't be player specific logic in entity if it can be handled by an override in Player (which keeps it self contained there)

I agree that instanceof is a codesmell but can we also agree that needing to pass in a position & relative bits for a method only used by player twice and other/custom entities way more is smelly as well?

instanceof is used in multiple places throughout the entity class and I think this is the lesser of two evils when considering other developers who will be extending entity in the future